### PR TITLE
TIEGCM initial instrument support

### PR DESCRIPTION
**Pull request by rstoneback**
Includes Instrument support for TIEGCM files from UCAR. All variables loaded into .data include 'time' as a dimension. Attributes and other small variables are attached to the Instrument object itself, per pysat design. I should check if that is documented....

I added TIEGCM to the supported instrument documentation.

I modified data setting to give xarray a chance at understanding multidimensional data before I do anything else.

I can't add models label for some reason... perhaps because it isn't my project.

### CODE REVIEW

**Comment by rstoneback** [comment]
Python 3 tests failing because of demeter

**Comment by aburrell** [comment]
> Python 3 tests failing because of demeter

Do you know where it's failing because of demeter?

**Comment by rstoneback** [comment]
There is a runtime error in downloads. Something about not being implemented yet.


On Oct 9, 2018, at 7:46 AM, Angeline Burrell <notifications@github.com<mailto:notifications@github.com>> wrote:


Python 3 tests failing because of demeter

Do you know where it's failing because of demeter?

**Comment by aburrell** [comment]
Right, I removed that Thurdsay so it shouldn't be a problem.  Must not have gotten pushed out.

**Comment by rstoneback** [comment]
Or I might not have pulled it in. My last push actually has a few commits in there, including leftovers from previous pushes. I could very well be out of sync.

My first reaction when seeing the feedback from Demeter was, I thought Angeline already fixed that up.


On Oct 9, 2018, at 7:52 AM, Angeline Burrell <notifications@github.com<mailto:notifications@github.com>> wrote:


Right, I removed that Thurdsay so it shouldn't be a problem. Must not have gotten pushed out.

**Comment by rstoneback** [comment]
I double checked the latest in develop. 

    url = 'https://cdpp-archive.cnes.fr/'
    print('Data must be downloaded by registered users at: {:s}'.format(url))
    raise RuntimeError('not written yet')

I'll pull in your changes when you have a chance to push. Cheers.

**Comment by aburrell** [comment]
Oh, I pushed the changes to model not develop.  I'll get to that today.

**Comment by aburrell** [comment]
I created a pull request with the DEMETER updates that I have completed at the moment (#92 )

File: pysat/_instrument.py [hunk 1]
```diff
@@ -534,12 +534,15 @@ def __setitem__(self, key, new):
                 elif len(np.shape(in_data)) == 0:
                     self.data[key] = ('time', [in_data]*len(self.index))
                 else:
-                    # multidimensional input
-                    # user needs to provide what is required
-                    if isinstance(in_data, tuple):
+                    try:
```
Comment (aburrell) [inline pysat/_instrument.py]: Is try/except the best way to do this?  What about turning it around:

`if isinstance(in_data, tuple):`
`    self.data[key] = in_data`
`else:`
`    try:`
`        self.data[key] = in_data`
`    except err:`
`        raise ValueError('unknown dimensions for data: {:}'.format(err))`

File: pysat/instruments/ucar_tiegcm.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+def list_files(tag=None, sat_id=None, data_path=None, format_str=None):
+    """Produce a list of files corresponding to UCAR TIEGCM.
+
+    This routine is invoked by pysat and is not intended for direct use by the end user.
+    
+    Multiple data levels may be supported via the 'tag' input string.
+    Currently defaults to level-2 data, or L2 in the filename.
+
+    Parameters
+    ----------
```
Comment (aburrell) [inline pysat/instruments/ucar_tiegcm.py]: Should specify default values

File: pysat/instruments/ucar_tiegcm.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    discarded. This routine uses the pysat.Files.from_os constructor, thus
+    the returned files are up to pysat specifications.
+    
+    """
+    
+    return pysat.Files.from_os(data_path=data_path, 
+                                format_str='tiegcm_icon_merg2.0_totTgcm.s_{day:03d}_{year:4d}.nc')
+
+def download(date_array, tag, sat_id, data_path=None, user=None, password=None):
+    print ('Not implemented.')
```
Comment (aburrell) [inline pysat/instruments/ucar_tiegcm.py]: Need docstring

**Comment by rstoneback** [comment]
I updated the default values. Added a docstring to downloads. I think I improved the xarray set item, I put the assignment under an isinstance check.

**Pull request merged**